### PR TITLE
FileTea

### PR DESCRIPTION
**Pull request by JosephKiranBabu**
Integration of FileTea service (#839) into FreedomBox. 

Sharing files is still difficult though we have integrated Syncthing (which is folder-sharing). A volatile file sharing service can make sharing files painless for FreedomBox users.

## Authentication
I have added Apache Basic Auth for the files shared via FileTea, which means that the receiver of the file should also be a Plinth user on the same FreedomBox. I am not sure of this decision. Can be convinced otherwise. The default permissions for FileTea is "anyone with the link can access". I turned it into "only logged in users of Plinth can access".

### CODE REVIEW

File: plinth/modules/filetea/__init__.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+def setup(helper, old_version=None):
+    """Install and configure the module."""
+    helper.install(managed_packages)
+    helper.call('configuration', actions.superuser_run, 'filetea',
+                ['configure'])
+    action_utils.service_restart('filetea')
+    helper.call('post', actions.superuser_run, 'filetea', ['enable'])
+    global service
+    if service is None:
```
Comment (jvalleroy) [inline plinth/modules/filetea/__init__.py]: I get an error: "name 'service' is not defined".

File: data/etc/apache2/conf-available/filetea-plinth.conf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Listen 8686
+
+<VirtualHost *:8686>
+      ProxyPass / http://localhost:8558/
+</VirtualHost>
```
Comment (jvalleroy) [inline data/etc/apache2/conf-available/filetea-plinth.conf]: Should apache's authentication be used here?

File: data/usr/lib/firewalld/services/filetea-plinth.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<service>
+    <short>FileTea</short>
+    <description>FileTea is an anonymous, volatile file storage system</description>
+    <port protocol="tcp" port="8558"/>
```
Comment (jvalleroy) [inline data/usr/lib/firewalld/services/filetea-plinth.xml]: Do both ports need to be opened in the firewall?

File: plinth/modules/filetea/__init__.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+description = [
+    _('FileTea is an anonymous, volatile file sharing service. '
+      'Volatile means the shared file is available only as long '
+      'as the sender keeps FileTea open on their browser.'),
+    format_lazy(
+        _('Running FileTea on {box_name} allows you to share files securely '
+          'with other users of {box_name}. On a public FileTea service, '
+          'anyone with the link can download a shared file. But FileTea on '
+          'FreedomBox is password protected, so only registered users of that '
+          'FreedomBox can download shared files, not any snooping third '
```
Comment (jvalleroy) [inline plinth/modules/filetea/__init__.py]: I think the next 2 uses of "FreedomBox" can be replaced with box_name.

File: plinth/modules/filetea/__init__.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    menu = main_menu.get('apps')
+    menu.add_urlname(title, 'glyphicon-share', 'filetea:index')
+
+    global service
+    setup_helper = globals()['setup_helper']
+    if setup_helper.get_state() != 'needs-setup':
+        service = service_module.Service(
+            managed_services[0],
+            title,
+            ports=['http', 'https'],
```
Comment (jvalleroy) [inline plinth/modules/filetea/__init__.py]: I think you will want to replace the ports here, with the service name "filetea-plinth".

File: plinth/modules/filetea/__init__.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    results.append(
+        action_utils.diagnose_url(
+            'http://localhost:8686', kind='4', check_certificate=False))
+    results.append(
+        action_utils.diagnose_url(
+            'http://localhost:8686', kind='6', check_certificate=False))
+    results.append(
+        action_utils.diagnose_url(
+            'https://{}:5678'.format(get_domainname()),
```
Comment (jvalleroy) [inline plinth/modules/filetea/__init__.py]: Should this be port 8686?

File: plinth/modules/filetea/urls.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+from django.conf.urls import url
+
+from plinth.views import ServiceView
+from plinth.modules import filetea
+
+urlpatterns = [
+    url(r'^apps/filetea/$',
+        ServiceView.as_view(
+            service_id=filetea.managed_services[0],
+            diagnostics_module_name='syncthing',
```
Comment (jvalleroy) [inline plinth/modules/filetea/urls.py]: syncthing -> filetea :)

**Comment by JosephKiranBabu** [comment]
@jvalleroy FileTea doesn't have an official logo. I took a screenshot of the title that appears on the FileTea page. It's not an image, just a h1 HTML tag. I am adding an empty license in the LICENSES file for now. Will get a logo designed for FileTea later.

**Comment by jvalleroy** [comment]
@JosephKiranBabu That sounds fine. The only remaining issue I see if the firewall service definition. With 'ports' set to 'http' and 'https', it will only open TCP ports 80 and 443 for this service. By switching the port to 'filetea-plinth', it will open the TCP port defined in the service file.

**Comment by JosephKiranBabu** [comment]
@jvalleroy I found that FileTea uses an older version of jQuery which makes it unusable on Debian unstable. I initially tried it out on Jessie where it worked. A bug has been raised on Debian for this.

**Comment by lispyclouds** [comment]
I've raised the bug here with the maintainer https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=862742

File: plinth/modules/filetea/__init__.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    helper.call('configuration', actions.superuser_run, 'filetea',
+                ['configure'])
+    action_utils.service_restart('filetea')
+    helper.call('post', actions.superuser_run, 'filetea', ['enable'])
+    global service
+    if service is None:
+        service = service_module.Service(
+            managed_services[0],
+            title,
+            ports=['http', 'https'],
```
Comment (jvalleroy) [inline plinth/modules/filetea/__init__.py]: Need to update these ports also.

File: data/etc/apache2/conf-available/filetea-plinth.conf [hunk 9]
```diff
@@ -0,0 +1,6 @@
+Listen 8686
+
+<VirtualHost *:8686>
+    Include includes/freedombox-auth-ldap.conf
```
Comment (jvalleroy) [inline data/etc/apache2/conf-available/filetea-plinth.conf]: I think this include can only be used in a Location:
```
May 18 23:39:05 freedombox apache2[4842]: The apache2 configtest failed. ... (warning).
May 18 23:39:05 freedombox apache2[4842]: Output of config test was:
May 18 23:39:05 freedombox apache2[4842]: AH00526: Syntax error on line 1 of /etc/apache2/includes/freedombox-auth-ldap.conf:
May 18 23:39:05 freedombox apache2[4842]: AuthType not allowed here
```

**Comment by lispyclouds** [comment]
@jvalleroy We are holding onto this until we have a fix for the aforementioned bug. That is pretty critical and the app is unusable on unstable.

**Comment by JosephKiranBabu** [comment]
Made the requested changes in ports. Will update the authentication config for apache once filetea becomes usable and testable.

**Comment by JosephKiranBabu** [comment]
@jvalleroy @lispyclouds 
The [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=862742) on the Debian package has been fixed but filetea's support for running behind a reverse proxy is available only from version 0.18.0.  See this [commit](https://github.com/elima/FileTea/commit/ccbcd6d6cce4ecde760c5a90a9d94e82d65b7019).
We should either explore an alternative solution meanwhile or wait for that version.

**Comment by JosephKiranBabu** [comment]
Created a bug on the Debian package for the upgrade to 0.1.18.
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=870597

**Comment by JosephKiranBabu** [comment]
The [bug on Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=870597) has stayed open for 2 months now. 
Maybe we should try to give a patch for that incompatibility issue with EventDance library.

Meanwhile, suggestions are welcome for a better solution to the file-sharing problem.

**Comment by JosephKiranBabu** [comment]
TODO:

Evaluate other alternatives for Filetea.
https://wiki.debian.org/FreedomBox/LeavingTheCloud

**Comment by JosephKiranBabu** [comment]
Though the problem of easily sharing a file still exists, FileTea has proved to be a badly maintained solution. 
I will consider integrating Coquelicot into Plinth. Coquelicot needs to upload the file to the FreedomBox server first, but otherwise is just as good as FileTea. 
We have to look at some other alternative for sharing big files anonymously (based on WebRTC maybe?). Our users still have the option of using BitTorrent for sharing big files non-anonymously (can be encrypted beforehand with a password emailed using GPG encryption).

**Pull request closed without merging**